pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -1840,94 +1840,123 @@ func TestPerPodSchedulingMetrics(t *testing.T) {
-	queue.Add(logger, pod)
-	pInfo, err = queue.Pop()
-	if err != nil {
-		t.Fatalf("Failed to pop a pod %v", err)
-	}
-	queue.AddUnschedulableIfNotPresent(logger, pInfo, 1)
-	// Override clock to exceed the DefaultPodMaxInUnschedulablePodsDuration so that unschedulable pods
-	// will be moved to activeQ
-	c.SetTime(timestamp.Add(DefaultPodMaxInUnschedulablePodsDuration + 1))
-	queue.flushUnschedulablePodsLeftover(logger)
-	pInfo, err = queue.Pop()
-	if err != nil {
-		t.Fatalf("Failed to pop a pod %v", err)
-	}
-	checkPerPodSchedulingMetrics("Attempt twice", t, pInfo, 2, timestamp)
+	tests := []struct {
+		name                            string
+		perPodSchedulingMetricsScenario func(*testingclock.FakeClock, *v1.Pod) (*framework.QueuedPodInfo, error)
+		wantAttempts                    int
+		wantInitialAttemptTs            time.Time
+	}{
+		{
+			// The queue operations are Add -> Pop.
+			name: "pod is created and scheduled after 1 attempt",
+			perPodSchedulingMetricsScenario: func(c *testingclock.FakeClock, pod *v1.Pod) (*framework.QueuedPodInfo, error) {
+				queue := NewTestQueue(ctx, newDefaultQueueSort(), WithClock(c))
+				queue.Add(logger, pod)
+				return queue.Pop()
+			},
+			wantAttempts:         1,
+			wantInitialAttemptTs: timestamp,
+		},
+		{
+			// The queue operations are Add -> Pop -> AddUnschedulableIfNotPresent -> flushUnschedulablePodsLeftover -> Pop.
+			name: "pod is created and scheduled after 2 attempts",
+			perPodSchedulingMetricsScenario: func(c *testingclock.FakeClock, pod *v1.Pod) (*framework.QueuedPodInfo, error) {
+				queue := NewTestQueue(ctx, newDefaultQueueSort(), WithClock(c))
+				queue.Add(logger, pod)
+				pInfo, err := queue.Pop()
+				if err != nil {
+					t.Fatalf("Failed to pop a pod %v", err)
+				}
 
-	// Case 3: Similar to case 2, but before the second pop, call update, the queue operations are
-	// Add -> Pop -> AddUnschedulableIfNotPresent -> flushUnschedulablePodsLeftover -> Update -> Pop.
-	c = testingclock.NewFakeClock(timestamp)
-	queue = NewTestQueue(ctx, newDefaultQueueSort(), WithClock(c))
-	queue.Add(logger, pod)
-	pInfo, err = queue.Pop()
-	if err != nil {
-		t.Fatalf("Failed to pop a pod %v", err)
-	}
-	queue.AddUnschedulableIfNotPresent(logger, pInfo, 1)
-	// Override clock to exceed the DefaultPodMaxInUnschedulablePodsDuration so that unschedulable pods
-	// will be moved to activeQ
-	c.SetTime(timestamp.Add(DefaultPodMaxInUnschedulablePodsDuration + 1))
-	queue.flushUnschedulablePodsLeftover(logger)
-	newPod := pod.DeepCopy()
-	newPod.Generation = 1
-	queue.Update(logger, pod, newPod)
-	pInfo, err = queue.Pop()
-	if err != nil {
-		t.Fatalf("Failed to pop a pod %v", err)
-	}
-	checkPerPodSchedulingMetrics("Attempt twice with update", t, pInfo, 2, timestamp)
+				queue.AddUnschedulableIfNotPresent(logger, pInfo, 1)
+				// Override clock to exceed the DefaultPodMaxInUnschedulablePodsDuration so that unschedulable pods
+				// will be moved to activeQ
+				updatedTimestamp := timestamp
```

Comment:
I don't think we need to make a copy, the Add below returns a copy with the updated timestamp, right?

Reply:
Addressed.